### PR TITLE
Implement parametrized rawquery (issue #115)

### PR DESCRIPTION
**Pull request by JohannesDaniel**

### CODE REVIEW

File: querqy-core/src/main/java/querqy/rewrite/commonrules/LineParser.java [hunk 1]
```diff
@@ -161,6 +170,32 @@ public static Object parse(final String line, final Input previousInput,
...
+        if (size == 1) {
+            return new StringRawQuery(null, rawQuery, occur, false);
+        } else {
+            final List<ParametrizedRawQuery.Part> parts = IntStream.range(0, size)
+                    .mapToObj(index -> index % 2 == 0
+                            ? new ParametrizedRawQuery.Part(rawQueryParts.get(index),
+                                ParametrizedRawQuery.Part.Type.QUERY_PART)
+                            : new ParametrizedRawQuery.Part(rawQueryParts.get(index).trim(),
+                                ParametrizedRawQuery.Part.Type.PARAMETER))
+                    .filter(part -> !part.isEmpty())
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/commonrules/LineParser.java:192]: I'm 99.9% sure we do not want empty parts but what if someone needs it. I guess it wouldn't hurt to leave them in?!

File: querqy-core/src/test/java/querqy/model/ParametrizedRawQueryTest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+                        new ParametrizedRawQuery.Part("q1 ", ParametrizedRawQuery.Part.Type.QUERY_PART),
+                        new ParametrizedRawQuery.Part("p1", ParametrizedRawQuery.Part.Type.PARAMETER),
+                        new ParametrizedRawQuery.Part(" q2 ", ParametrizedRawQuery.Part.Type.QUERY_PART),
+                        new ParametrizedRawQuery.Part("p2", ParametrizedRawQuery.Part.Type.PARAMETER)
+                ),
+                Clause.Occur.SHOULD,
+                false);
+
+        String queryString = rq.buildQueryString(param -> param.replace("p", "param"));
+        assertThat(queryString).isEqualTo("q1 param1 q2 param2");
```
Comment (renekrie) [inline querqy-core/src/test/java/querqy/model/ParametrizedRawQueryTest.java:25]: I think there shouldn't be a space - except it's part of QUERY_PART or PARAM.

**Review by renekrie — CHANGES_REQUESTED** [review]
Thank you for this great effort!

I'm afraid I disagree about how to handle spaces. For robustness, I would make no assumption about the meaning of spaces in the underlying query language (except for leading/trailing spaces of the full query string in our rules file). This means that I would neither trim parts of the query string nor inject a space when applying the parameters.

**Comment by JohannesDaniel** [comment]
How do you thing about spaces in the context of using the templates? SHould we support this?
`<< template_name: param1 = value1 || param2 = value2 >>`?

I think allowing spaces here and trimming values / params helps for better readability.

**Pull request merged**